tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuner.scala

```diff
@@ -0,0 +1,435 @@
+  }
+
+  def set_max_partition_bytes(maxPartitionBytes: String): Unit = {
+    properties("spark.sql.files.maxPartitionBytes") = maxPartitionBytes
+  }
+
+  def set_pinned_pool_size(pinnedPoolSize: String): Unit = {
+    properties("spark.rapids.memory.pinnedPool.size") = pinnedPoolSize
+  }
+
+  def set_executor_memory_overhead(executorMemoryOverhead: String): Unit = {
+    properties("spark.executor.memoryOverhead") = executorMemoryOverhead
+  }
+
+  def set_executor_memory_overhead_factor(executorMemoryOverheadFactor: Double): Unit = {
+    properties("spark.executor.memoryOverheadFactor") = executorMemoryOverheadFactor.toString
+  }
+}
+
+class AutoTuner(app: ApplicationSummaryInfo, workerInfo: String) extends Logging {
+  import AutoTuner._
+  val DEFAULT_SHUFFLE_PARTITION_MULTIPLIER: Int = 2
+  val MAX_JVM_GCTIME_FRACTION: Double = 0.3
+
+  val DEFAULT_CONCURRENT_GPU_TASKS_MULTIPLIER: Double = 0.125 // Currently aggressively set to 1/8
+  val MAX_CONCURRENT_GPU_TASKS: Int = 4
+
+  val DEFAULT_MAX_PARTITION_BYTES: String = "512m"
+  val MAX_PARTITION_BYTES_BOUND: String = "4g"
+  val MAX_PARTITION_BYTES_RANGE: String = "256m"
+  val MIN_PARTITION_BYTES_RANGE: String = "128m"
+
+  val DEFAULT_PINNED_POOL_SIZE: String = "2g"
+  val DEFAULT_MEMORY_OVERHEAD_FACTOR: Double = 0.1
+  val MIN_MEMORY_OVERHEAD = "2g"
+  val MAX_EXTRA_SYSTEM_MEMORY = "2g"
+
+  val MAX_PER_EXECUTOR_CORE_COUNT: Int = 16
+  val MIN_PER_EXECUTOR_CORE_COUNT: Int = 4
+
+  val MAX_EXECUTOR_MEMORY: String = "64g"
+  val MIN_EXECUTOR_MEMORY: String = "8g"
+
+  var comments: Seq[String] = Seq()
+
+  private def recommendSparkProperties(bestConfig: Config, systemProps: SystemProps): Unit = {
+    if (systemProps == null) {
+      logWarning("System information is not available. Cannot recommend properties.")
+      comments :+= "'spark.executor.memory' should be set to at least 2GB/core."
+      comments :+= "'spark.executor.instances' should be set to 'num_gpus * num_workers'."
+    } else {
+      // Recommendation for 'spark.executor.instances' based on number of gpus and workers
+      systemProps.num_workers match {
+        case Some(numWorkers) =>
+          val numInstances = if (systemProps.gpu_props != null) {
+            numWorkers * systemProps.gpu_props.count
+          } else numWorkers
+          bestConfig.set_executor_instances(numInstances)
+        case None =>
+          val num_gpus_str = if (systemProps.gpu_props != null) {
+            systemProps.gpu_props.count.toString
+          } else "num_gpus"
+
+          comments :+= s"'spark.executor.instances' should be set to $num_gpus_str * num_workers."
+      }
+
+      // Recommendation for 'spark.executor.cores' based on number of cpu cores and gpus
+      val numCores: Int = if (systemProps.gpu_props != null) {
+        Math.min(systemProps.numCores * 1.0 / systemProps.gpu_props.count,
+          MAX_PER_EXECUTOR_CORE_COUNT).toInt
+      } else systemProps.numCores
```

Comment:
if we are assuming this is alwasy set and accurate we need to make sure file in mandatory and lets user know.

Reply:
note we may be able to get some of the information from the profiling tool about the executors if they don't run the script